DigitalLearningSolutions.Web/Scripts/searchSortFilterAndPaginate/filter.ts

```diff
@@ -0,0 +1,253 @@
+
+export interface IAppliedFilter {
+  group: string;
+  propertyName: string;
+  propertyValue: string;
+  filterValue: string;
+}
+
+export interface IAppliedFilterTag {
+  element: Element;
+  filterValue: string;
+}
+
+export const separator = '|';
+export const filterSeparator = '╡';
+
+export function setUpFilter(onFilterUpdated: VoidFunction): void {
+  setUpFilterSubmitButtons();
+  setUpClearFiltersButton();
+  setUpFilterSelectorDropdown();
+  hideAllFilterDropdowns();
+  document.getElementById('filter-by')?.addEventListener('change', onFilterUpdated);
+}
+
+export function filterSearchableElements(
+  searchableElements: ISearchableElement[],
+  possibleFilters: IAppliedFilterTag[],
+): ISearchableElement[] {
+  let filteredSearchableElements = searchableElements;
+  const filterBy = getFilterBy();
+  const appliedFilters = getAppliedFilters(filterBy);
+
+  const filterGroups = _.groupBy(appliedFilters, (filter) => filter.group);
+
+  _.forEach(filterGroups, (appliedFiltersInGroup) => {
+    const itemsToFilter = filteredSearchableElements;
+    const setOfFilteredLists = _.map(
+      appliedFiltersInGroup,
+      (af) => filterElements(itemsToFilter, af),
+    );
+    const flattenedElements = _.flatten(setOfFilteredLists);
+    filteredSearchableElements = _.uniq(flattenedElements);
+  });
+
+  if (filterBy) {
+    updateAppliedFilters(filterBy, possibleFilters);
+    showAppliedFilters();
+  }
+
+  return filteredSearchableElements;
+}
+
+function setUpFilterSubmitButtons() {
+  const filterSubmits = Array.from(
+    document.getElementsByClassName('filter-submit__dropdown'),
+  );
+  filterSubmits.forEach((filterSubmit) => {
+    const element = <HTMLInputElement>filterSubmit;
+    element.addEventListener('click', (e) => {
+      e.preventDefault();
+      const newFilter = appendNewFilterToFilterBy(element);
+      if (newFilter) {
+        hideAllFilterDropdowns();
+        resetFilterSelectorDropdown();
+      }
+    });
+  });
+}
+
+function setUpClearFiltersButton() {
+  document.getElementById('clear-filters')?.
+    addEventListener('click', (e) => {
+      e.preventDefault();
+      clearFilters();
+    });
+}
+
+function setUpFilterSelectorDropdown() {
+  document.getElementById('filter-selector')?.
+    addEventListener('change', () => {
+      showSelectedFilterDropdown();
+    });
+}
+
+function getAppliedFilters(filterBy: string): IAppliedFilter[] {
+  return filterBy.split(filterSeparator)
+    .map((filter) => newAppliedFilterFromFilter(filter));
+}
+
+function newAppliedFilterFromFilter(filter: string): IAppliedFilter {
+  const splitFilter = filter.split(separator);
+  return <IAppliedFilter>{
+    group: splitFilter[0],
+    propertyName: splitFilter[1],
+    propertyValue: splitFilter[2],
+    filterValue: filter,
+  };
+}
+
+function filterElements(
+  searchableElements: ISearchableElement[],
+  appliedFilter: IAppliedFilter,
+): ISearchableElement[] {
+  return _.filter<ISearchableElement>(

```

Comment:
This is minor, but could we just use JS native filter (as the syntax is a bit nicer) & for find on L236